GH-43809: [Docs] Update extension type examples to not use UUID (#43849)
apache/arrow: docs/source/python/extending_types.rst

Defining extension types ("user-defined types")
-----------------------------------------------

Arrow affords a notion of extension types which allow users to annotate data

> types with additional semantics. This allows downstream consumers both to

Reviewer: I'm -1 on the "downstream consumers" phrasing. How about just "developers"?

Author: Fair. I had in mind that end users might also come to rely on metadata (e.g., you might use this extension type to create an SPSS-like survey data format), but this seems like a reasonably niche use case.

specify custom serialization and deserialization routines (for example,
to :ref:`Python scalars <custom-scalar-conversion>` and
:ref:`pandas <conversion-to-pandas>`) and to more easily interpret data.

> In the Arrow :doc:`metadata specification</format/Columnar.html#format-metadata-extension-types>`,

Reviewer: Please use the existing ref. 
```suggestion
In the Arrow :ref:`format_metadata_extension_types`,
```

This points here:
https://github.com/apache/arrow/blob/ab0a40ee34217070f14027776682074c55d0b507/docs/source/format/Columnar.rst?plain=1#L1575-L1578

Author: Thanks! Missed that reference.

this is accomplished by annotating any of the built-in Arrow data types
(the "storage type") with a custom type name and, optionally, a byte
array that can be used to provide additional metadata (referred to as
"parameters" in this documentation). These appear as the
``ARROW:extension:name`` and ``ARROW:extension:metadata`` keys in the
Field's ``custom_metadata``.

Note that since these annotations are part of the Arrow specification,
they can potentially be recognized by other (non-Python) Arrow consumers
such as PySpark.

> Pyarrow allows you to define extension types from Python by subclassing

Reviewer: ```suggestion
PyArrow allows you to define extension types from Python by subclassing
```

can potentially be recognized by other (non-Python) Arrow implementations
such as PySpark.

For example, we could define a custom rational type for fractions which can
be represented as a pair of integers::

    import pyarrow as pa
    import pyarrow.types as pt

> 

Reviewer: I think it's better to:
- omit these imports from the example
- just use `pa.types` in the code (instead of `pt`) here and in the other files

Author: Done

    class RationalType(pa.ExtensionType):

        def __init__(self):

            super().__init__(
                pa.struct(
                    [
                        ("numer", pa.int32()),
                        ("denom", pa.int32()),
                    ],
                ),
                "my_package.rational",
            )

        def __arrow_ext_serialize__(self) -> bytes:
            # No serialized metadata necessary
            return b""

        @classmethod

>         def __arrow_ext_deserialize__(self, storage_type, serialized):

Reviewer: ```suggestion
        def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

            # Sanity checks, not required but illustrate the method signature.
            assert pt.is_struct(storage_type)

>             assert pt.is_int32(storage_type[0].type)

Reviewer: Might as well check denom too
```suggestion
            assert pt.is_int32(storage_type[0].type)
            assert pt.is_int32(storage_type[1].type)
```

>             assert serialized == b''

Reviewer: ```suggestion
            assert serialized == b""
```

Author: A question: There are several other places in this file where single quotes were used, e.g., https://github.com/apache/arrow/blob/fde32158fa5f9a5f943da5aacd2bd8056b7a5bb1/docs/source/python/extending_types.rst?plain=1#L267

Would it be helpful to add an issue to standardize to double quotes?

Reviewer: > Would it be helpful to add an issue to standardize to double quotes?

Sure, or if you just want to fix them in this PR, that's fine too. I don't think we're too fussy about consistency of single/double quotes in examples, but it's best to at least be consistent within one example.

Author: Fair! I'll go ahead and change them all while I'm here.

            return b""

        @classmethod
        def __arrow_ext_deserialize__(cls, storage_type, serialized):
        def __arrow_ext_deserialize__(self, storage_type, serialized):
            # Sanity checks, not required but illustrate the method signature.
            assert storage_type == pa.binary(16)
            assert pt.is_struct(storage_type)
            assert pt.is_int32(storage_type[0].type)
            assert serialized == b''
            # Return an instance of this subclass given the serialized
            # metadata.
            return UuidType()

            # return an instance of this subclass given the serialized
            # metadata
            return RationalType()

The special methods ``__arrow_ext_serialize__`` and ``__arrow_ext_deserialize__``
define the serialization of an extension type instance. For non-parametric

> types such as the above, the serialization payload can be left empty.

Reviewer: Why remove this?

Reviewer: Oh, I see. I think this relates to what exactly parameterized / parametric means, correct?

- If the storage type requires one or more parameters to initially specify, does that qualify as "parameterized"?
- Or does "parameterized" mean that we need additional metadata (that is not encoded in the underlying storage type) in order to reconstruct the instance?

I think you are defining it to mean the latter, correct?

I think that makes sense. But it would be good to explicitly say this in the docs. See my other comment about that.

>     ... type=rational_type.storage_type

Reviewer: Please indent (like in the example in `types.pxi` below)

Author: Done!

    ... )

>     >>> arr = rational_type.wrap_array(storage_array)
>     >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)

Reviewer: ```suggestion
    >>> arr = rational_type.wrap_array(storage_array)
    # or equivalently
    >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
```

Author: This is the same as the original documentation. I'm guessing the previous writer didn't insert `# or equivalently` as as it causes `doctest` to fail. I've changed this to the following:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)
>>> # or equivalently
>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

But the alternative (which I think is harder to read) is:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)

or equivalently

>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

Reviewer: Thanks. That looks good to me.

    >>> with pa.ipc.open_stream(buf) as reader:
    ...    result = reader.read_all()
    >>> result.column("ext").type
    RationalType(StructType(struct<numer: int32, denom: int32>))

Further, note that while we registered the concrete type

> ``RationalType(pa.int32())``, ``RationalType(integer_type)`` has the same
> extension name (``"my_package.rational"``) for all integer types. As such,
> the above code will also allow users to (de)serialize these data types::

Reviewer: Edit for readability
```suggestion
``RationalType(pa.int32())``, you could use the same extension name
``"my_package.rational"`` to register ``RationalType(integer_type)`` for
any Arrow integer type. As such, the above code will also allow users to
(de)serialize these data types::
```

Author: Adopted with a small change.

            # attributes need to be set first before calling
            # super init (as that calls serialize)
            self._freq = freq
            super().__init__(pa.int64(), "my_package.period")

        @property
        def freq(self):
            return self._freq

        def __arrow_ext_serialize__(self):

>             return "freq={self.freq}".encode()

Reviewer: Please revert the change to this one line. It was correct before but now it is not.
